os/include/tinyara/usb/usbhost.h

```diff
@@ -587,21 +681,74 @@ struct usbhost_registry_s {
 	FAR const struct usbhost_id_s *id;	/* An array of ID info. Actual dimension is nids */
 };
 
+/* This type represents one endpoint configured by the epalloc() method.
+ * The actual form is known only internally to the USB host controller
+ * (for example, for an OHCI driver, this would probably be a pointer
+ * to an endpoint descriptor).
+ */
+
+typedef FAR void *usbhost_ep_t;
+/* In the hierarchy of things, there is the host control driver (HCD),
+ * represented by struct usbhost_driver_s.  Connected to the HCD are one
+ * or more hubs.  At a minimum, the root hub is always present. Each hub
+ * has from 1 to 4 ports.
+ */
+
+/* Every class connects to the host controller driver (HCD) via a port on a
+ * hub.  That hub may be an external hub or the internal, root hub.  The
+ * root hub is managed by the HCD.  This structure describes that state of
+ * that port and provides the linkage to the parent hub in that event that
+ * the port is on an external hub.
+ *
+ * The root hub port can be distinguish because it has parent == NULL.
+ */
+struct usbhost_hubport_s {
+	FAR struct usbhost_driver_s *drvr;	/* Common host driver */
+#ifdef CONFIG_USBHOST_HUB
+	FAR struct usbhost_hubport_s *parent;	/* Parent hub (NULL=root hub) */
+#endif
+	FAR struct usbhost_class_s *devclass;	/* The bound device class driver */
+	usbhost_ep_t ep0;			/* Control endpoint, ep0 */
+	bool connected;				/* True: device connected; false: disconnected */
+	uint8_t port;				/* Hub port index */
+	uint8_t funcaddr;			/* Device function address */
+	uint8_t speed;				/* Device speed */
+};
+
+/* The root hub port differs in that it includes a data set that is used to
+ * manage the generation of unique device function addresses on all
+ * downstream ports.
+ */
+
+struct usbhost_roothubport_s {
+	/* This structure must appear first so that this structure is cast-
+	 * compatible with usbhost_hubport_s.
+	 */
+
+	struct usbhost_hubport_s hport;	/* Common hub port definitions */
+	struct usbhost_devaddr_s devgen;	/* Address generation data */
+};
 /* struct usbhost_class_s provides access from the USB host driver to the USB host
  * class implementation.
  */
 
 struct usbhost_class_s {
+	/* Class instances are associated with devices connected on one port on a
+	 * hub and are represented by this structure.
+	 */
+
+	FAR struct usbhost_hubport_s *hport;	/* The port used by this class instance */
+
 	/* Provides the configuration descriptor to the class.  The configuration
 	 * descriptor contains critical information needed by the class in order to
 	 * initialize properly (such as endpoint selections).
 	 */
 
-	int (*connect)(FAR struct usbhost_class_s *usbclass, FAR const uint8_t *configdesc, int desclen, uint8_t funcaddr);
+	int (*connect)(FAR struct usbhost_class_s *devclass, FAR const uint8_t *configdesc, int desclen /*, uint8_t funcaddr */);
```

Comment:
If funcaddr parameter is not needed can't we remove it fully.

Reply:
Removed